File: old/sorting/insertion_sort/count_insertion_shifts/shifts_so_far_insertion_sort.cpp
#include <iostream>
#include <vector>


using namespace std;

int countShiftsInsertionSort(vector<int>& arr) {
    int shiftSoFar = 0;
    for (int i = 1; i < arr.size(); i++) {
        int prevIndex = i - 1;
        int currVal = arr[i];
        while (prevIndex >= 0 && arr[prevIndex] > currVal) {
            arr[prevIndex + 1] = arr[prevIndex];
            prevIndex -= 1;
            shiftSoFar++;
        }
        arr[prevIndex + 1] = currVal;
    }
    return shiftSoFar;
}

int main() {
    vector<int> v = {3, 2, 1, 4, 5};   
    int shifts =  countShiftsInsertionSort(v);
    cout << shifts;
}

File: old/sorting/insertion_sort/count_insertion_shifts/shifts_so_far_insertion_sort.py
def shifts_so_far_insertion_sort(arr: list[int]) -> int:
    shifts_so_far = 0
    for i in range(1, len(arr)):
        prev_index = i - 1
        curr_val = arr[i]
        while (prev_index >= 0 and arr[prev_index] > curr_val):
            arr[prev_index + 1] = arr[prev_index]
            prev_index -= 1
            shifts_so_far += 1
        arr[prev_index + 1] = curr_val
    return shifts_so_far

def main():
    shifts = shifts_so_far_insertion_sort([3, 2, 1, 5, 4])
    print(shifts)

main()
